### PR TITLE
[CBRD-23260] Lock-free transaction system & table

### PR DESCRIPTION
**Pull request by ribram**
http://jira.cubrid.org/browse/CBRD-23260

- created lockfree::tran::system to manage indexes; this will allow multiple possible systems, each with its own tables\*
- thread manager has its own lockfree::tran::system; this would be used by all existing hashes.
- moved index/id into a definition header
- created lockree::tran::descriptor and lockfree::tran::table classes; added only basic start/end transaction functionality and management of global/min transaction ids.

\* let's say we will have a lock-free hash that will be used only for index parallel loading. using the same system as threads in thread manager would allocate many unnecessary transaction descriptor. better to make a separate system specifically for these threads.

TODO later:

  - rename lockfree_transaction_index files as lockfree_transaction_system
  - separate lockfree_transaction into lockfree_transaction_descriptor and lockfree_transaction_table
  - add a class for transactionable item (or hazard pointer, as @radudoros suggested), see https://en.wikipedia.org/wiki/Hazard_pointer
  - extensive unit testing

### CODE REVIEW

File: src/base/lockfree_transaction_index.hpp [hunk 1]
```diff
@@ -18,29 +18,52 @@
...
+	~system ();
+
+	index assign_index ();
+	void free_index (index idx);
+	size_t get_max_transaction_count () const;
+
+      private:
+	size_t m_max_tran_per_table;
+	std::mutex m_tran_idx_lock;
+	bitmap *m_tran_idx_map;
```
Comment (radudoros) [inline src/base/lockfree_transaction_index.hpp]: simple composition seems better than pimpl here

File: src/base/area_alloc.h [hunk 2]
```diff
@@ -29,11 +29,16 @@
...
+#error C++ is required
+#endif
+
 #if !defined (SERVER_MODE)
 #else /* SERVER_MODE */
 #include "connection_defs.h"
 #endif /* SERVER_MODE */
-#include "lock_free.h"
+#include "lockfree_bitmap.hpp"
+#include "porting.h"
```
[the remark this answers is not in the document]
Comment (ribram) [reply src/base/area_alloc.h]: imo it is better to respect the alphabetical order regardless.

any header that requires `porting.h` should include it.

File: src/base/lockfree_transaction_index.hpp (hunk 1 above)
Comment (ribram) [reply src/base/lockfree_transaction_index.hpp]: done

**Comment by ribram** [comment]
@eseokoh
I am merging this to unblock follow-ups. If you have a chance to review later and you want to request changes, I will make them in separate patches.

**Pull request merged**